### PR TITLE
Gpio cleanup

### PR DESCRIPTION
**Pull request by bugadani**
The next part of our journey, slightly simplifying the trait spaghetti behind GPIOs. Also marking all internal methods with `#[doc(hidden)]` to reduce the temptation.

### CODE REVIEW

File: esp-hal/src/gpio/mod.rs [hunk 1]
```diff
@@ -1044,57 +1033,36 @@ fn on_pin_irq(pin_nr: u8) {
...
-}
-
 #[doc(hidden)]
 #[macro_export]
 macro_rules! if_output_pin {
-    (InputOnlyAnalog, { $($then:tt)* } else { $($else:tt)* } ) => { $($else)* };
     (InputOutputAnalog, { $($then:tt)* } else { $($else:tt)* } ) => { $($then)* };
     (InputOutputAnalogTouch, { $($then:tt)* } else { $($else:tt)* } ) => { $($then)* };
     (InputOutput, { $($then:tt)* } else { $($else:tt)* } ) => { $($then)* };
+    ($other:ident, { $($then:tt)* } else { $($else:tt)* } ) => { $($else)* };
```
Comment (bugadani) [inline esp-hal/src/gpio/mod.rs:1042]: This macro didn't handle `InputOnly` before, even though we don't have any such pin.

File: esp-hal/src/gpio/mod.rs [hunk 2]
```diff
@@ -1044,57 +1033,36 @@ fn on_pin_irq(pin_nr: u8) {
...
 pub(crate) use if_output_pin;
 
 #[doc(hidden)]
 #[macro_export]
-macro_rules! pin_types {
-    (InputOnly) => {
-        type IsOutput = $crate::gpio::False;
-        type IsAnalog = $crate::gpio::False;
-        type IsTouch = $crate::gpio::False;
+macro_rules! io_types {
```
Comment (bugadani) [inline esp-hal/src/gpio/mod.rs:1048]: I guess the rename doesn't align with what I want to do long term, but whatever. Long term I want to declare all pin type traits (Analog, Touch, LP...) here, in such a way that we can move the per-chip implementation to the soc module where they belong.

**Comment by Dav1dde** [comment]
As a mere user, having internal stuff not be visible in the docs is a huge help especially for a beginner like me. This means stuff won't be auto completed I am not supposed to use and the docs search doesn't show a bunch of stuff you're not supposed to use.

**Review by bjoernQ — APPROVED** [review]
Good to see GpioProperties go away

**Review by MabezDev — APPROVED** [review]
This cleans up the docs considerably, thanks!

**Pull request merged**